### PR TITLE
Fleming design model: flow tests for merges

### PR DESCRIPTION
**Pull request by octol**
Add merges to CheckAndProcessElderChange flow tests

Render of graphs used: https://raw.githack.com/maidsafe/routing/bd2d7d32cb1fcffc57e652df3fa5ff0f92f430fe/routing_model/index.html

The sub-process ProcessMerge is missing and would be the next logical step

### CODE REVIEW

File: routing_model/src/flows_dst.rs [hunk 1]
```diff
@@ -278,7 +280,15 @@ impl CheckAndProcessElderChange {
 
     fn try_consensus(&self, vote: &ParsecVote) -> Option<Self> {
         if ParsecVote::CheckElder == *vote {
```
Comment (jeanphilippeD) [inline routing_model/src/flows_dst.rs]: I think you'd be better off extracting the ProcessElderChange sub routine first.
You can see an example of sub routine in the source or destination flows, but feel free to implement the sub-routine differently if it make sense (these other sub routine will be removed to match the new graph, but the elder/merge/split will remain).

Comment (jeanphilippeD) [reply routing_model/src/flows_dst.rs]: This would allow you to refactor and all the test should still pass.

File: routing_model/src/flows_dst.rs [hunk 2]
```diff
@@ -278,7 +280,15 @@ impl CheckAndProcessElderChange {
...
-            return Some(self.check_elder());
+            // TODO: something seems strange here
+            if self.has_merge_infos() {
+                // TODO delegate to ProcessMerge flow
+                return None;
+            } else {
+                return Some(self.check_elder());
+            }
+        } else if let ParsecVote::NeighbourMerge(info) = *vote {
```
Comment (jeanphilippeD) [inline routing_model/src/flows_dst.rs]: you'll want to use a match here.

File: routing_model/src/utilities.rs [hunk 3]
```diff
@@ -74,6 +74,9 @@ pub struct SectionInfo(pub Section, pub i32 /*contain full membership */);
 #[derive(Debug, Clone, Copy, Default, PartialEq, PartialOrd, Ord, Eq)]
 pub struct GenesisPfxInfo(pub SectionInfo);
 
+#[derive(Debug, Clone, Copy, Default, PartialEq, PartialOrd, Ord, Eq)]
+pub struct NeighbourMergeInfo;
```
Comment (jeanphilippeD) [inline routing_model/src/utilities.rs]: Maybe call it MergeInfo, because it comes from the Merge RPC

File: routing_model/src/flows_dst.rs [hunk 4]
```diff
@@ -343,6 +360,23 @@ impl CheckAndProcessElderChange {
...
+    fn vote_parsec_neighbour_merge(&self) -> Self {
+        self.0
+            .action
+            .vote_parsec(ParsecVote::NeighbourMerge(NeighbourMergeInfo));
+        self.clone()
+    }
+
+    fn store_merge_infos(&self, info: NeighbourMergeInfo) -> Self {
+        let mut state = self.clone();
+        state.mut_routine_state().neighbour_merge_info = Some(info);
```
Comment (jeanphilippeD) [inline routing_model/src/flows_dst.rs]: No, this is a function in the graph.
What this mean is that it needs to be implemented by the Action component.
Only local variable that appear in the graph as SCREAMING_CAMEL_CASE should be stored locally, or any local state you find yourself needing.

File: routing_model/src/scenario_tests.rs [hunk 5]
```diff
@@ -175,6 +176,7 @@ fn run_test(test_name: &str, start_state: &State, events: &[Event], expected_sta
         },
         final_state.failure,
     );
+    dbg!(&final_state);
```
Comment (jeanphilippeD) [inline routing_model/src/scenario_tests.rs]: Why did you need this? did the assert_eq not provide you the needed info?
If so, we should consider making sure it does.

File: routing_model/src/scenario_tests.rs [hunk 6]
```diff
@@ -564,6 +566,65 @@ mod dst_tests {
         );
     }
 
+    #[test]
+    fn test_parsec_expect_candidate_and_online_then_neighbour_merge() {
+        let initial_state = arrange_initial_state(
+            &initial_state_old_elders(),
+            &[
+                ParsecVote::ExpectCandidate(CANDIDATE_1).to_event(),
+                ParsecVote::Online(CANDIDATE_1).to_event(),
```
Comment (jeanphilippeD) [inline routing_model/src/scenario_tests.rs]: Why Online? It should only be used for resource proofing, I don't see it would do anything.

File: routing_model/src/scenario_tests.rs [hunk 7]
```diff
@@ -564,6 +566,65 @@ mod dst_tests {
...
+    // JON
+    #[test]
+    fn test_parsec_neighbour_merge() {
+        run_test(
+            "Get Parsec NeighbourMergeInfo",
+            &initial_state_old_elders(),
+            &[ParsecVote::NeighbourMerge(NeighbourMergeInfo).to_event()],
+            &AssertState {
+                check_and_process_elder_change_routine: CheckAndProcessElderChangeState {
+                    neighbour_merge_info: Some(NeighbourMergeInfo),
```
Comment (jeanphilippeD) [inline routing_model/src/scenario_tests.rs]: Because it is a store thing, for testing, you could also enqueu it in the our_events vector in InnerAction.
You'll also actually want to store it so you can get it back out when needed.

File: routing_model/src/state.rs [hunk 8]
```diff
@@ -18,6 +18,7 @@ use crate::utilities::*;
 pub struct CheckAndProcessElderChangeState {
     pub wait_votes: Vec<ParsecVote>,
     pub change_elder: Option<ChangeElder>,
+    pub neighbour_merge_info: Option<NeighbourMergeInfo>,
```
Comment (jeanphilippeD) [inline routing_model/src/state.rs]: As mentioned, should not be there.
You see wait_votes as a couter part WAIT_VOTES.
and change_elder is helpful to implement the update to elder, but is not something that appear in the diagram at all.

File: routing_model/src/scenario_tests.rs (hunk 5 above)
Comment (octol) [reply routing_model/src/scenario_tests.rs]: Well it's the same info, but dbg gives it with linebreaks and indentation. I struggle to parse it when it's just one long long line :(

File: routing_model/src/scenario_tests.rs (hunk 6 above)
Comment (octol) [reply routing_model/src/scenario_tests.rs]: Yes I think this was actually just a placeholder

File: routing_model/src/utilities.rs (hunk 3 above)
Comment (octol) [reply routing_model/src/utilities.rs]: Done

File: routing_model/src/flows_dst.rs (hunk 4 above)
Comment (octol) [reply routing_model/src/flows_dst.rs]: Ah yes good point!

File: routing_model/src/scenario_tests.rs (hunk 5 above)
Comment (jeanphilippeD) [reply routing_model/src/scenario_tests.rs]: https://crates.io/crates/pretty-assertions ?

File: routing_model/src/flows_dst.rs [hunk 9]
```diff
@@ -281,11 +284,41 @@ impl CheckAndProcessElderChange {
...
         }
     }
 
+    fn store_merge_infos(&self, merge_info: MergeInfo) -> Self {
+        self.0.action.store_merge_infos(merge_info);
+        self.clone()
+    }
+
+    fn merge_needed(&self) -> bool {
+        // TODO: trigger for merge needed still in flux
```
Comment (jeanphilippeD) [inline routing_model/src/flows_dst.rs]: For this PR, we would want to have this return true and false.

You can choose a criteria that is easy to implement and somewhat sensible, it does not have to relate to what the real merge needed require: I would say for now, just mock it (i.e have a field in InnerAction: merge_needed:bool

**Comment by pierrechevalier83** [comment]
We have moved the routing model to its own repo instead of a branch in routing.
This PR was moved to https://github.com/maidsafe/routing_model/pull/1
